src/passes/GlobalEffects.cpp

```diff
@@ -22,60 +22,184 @@
 #include "ir/effects.h"
 #include "ir/module-utils.h"
 #include "pass.h"
+#include "support/unique_deferring_queue.h"
 #include "wasm.h"
 
 namespace wasm {
 
 struct GenerateGlobalEffects : public Pass {
   void run(Module* module) override {
-    // TODO: Full transitive closure of effects. For now, we just look at each
-    //       function by itself.
+    // First, we do a scan of each function to see what effects they have,
+    // including which functions they call directly (so that we can compute
+    // transitive effects later).
 
-    auto& funcEffectsMap = getPassOptions().funcEffectsMap;
-
-    // First, clear any previous effects.
-    funcEffectsMap.reset();
+    struct FuncInfo {
+      // Effects in this function.
+      std::unique_ptr<EffectAnalyzer> effects;
```

Comment:
Would it make sense to use `std::optional` here?

Reply:
Good idea, done.

I actually didn't know how to do late construction with optional, but it turns out you can do `optional.emplace(..)` for that...